File: solutions/3100-3199/3123.find-edges-in-shortest-paths/Solution.cpp
class Solution
{
public:
  vector<bool> findAnswer(int n, vector<vector<int>> &edges)
  {
    vector<bool> res(edges.size()), visited(n);
    vector<vector<array<int, 2>>> al(n);
    for (int i = 0; i < edges.size(); ++i)
    {
      al[edges[i][0]].push_back({edges[i][1], i}); // track edge index so we
      al[edges[i][1]].push_back({edges[i][0], i}); // we can update the result
    }
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;
    vector<int> path(n, INT_MAX);
    pq.push({0, 0});
    path[0] = 0;
    while (!pq.empty())
    {
      auto [len, i] = pq.top();
      pq.pop();
      if (len == path[i])
        for (auto [j, e] : al[i])
          if (len + edges[e][2] < path[j])
          {
            path[j] = len + edges[e][2];
            pq.push({path[j], j});
          }
    }
    pq.push({path[n - 1], n - 1}); // backtrack
    while (!pq.empty())
    {
      auto [len, i] = pq.top();
      pq.pop();
      for (auto [j, e] : al[i])
        if (len - edges[e][2] == path[j])
        {
          if (!visited[j])
            pq.push({path[j], j});
          res[e] = visited[j] = true;
        }
    }
    return res;
  }
};

File: solutions/3100-3199/3123.find-edges-in-shortest-paths/Solution.java
import java.util.*;

class Solution {
    public boolean[] findAnswer(int n, int[][] edges) {
        boolean[] res = new boolean[edges.length];
        boolean[] visited = new boolean[n];

        List<List<int[]>> al = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            al.add(new ArrayList<>());
        }

        for (int i = 0; i < edges.length; i++) {
            int[] edge = edges[i];
            al.get(edge[0]).add(new int[]{edge[1], i});
            al.get(edge[1]).add(new int[]{edge[0], i});
        }

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        int[] dist = new int[n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        pq.offer(new int[]{0, 0});
        dist[0] = 0;

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int d = current[0];
            int u = current[1];

            if (d > dist[u]) continue;

            for (int[] neighbor : al.get(u)) {
                int v = neighbor[0];
                int e = neighbor[1];
                int newDist = d + edges[e][2];
                if (newDist < dist[v]) {
                    dist[v] = newDist;
                    pq.offer(new int[]{dist[v], v});
                }
            }
        }

        if (dist[n - 1] == Integer.MAX_VALUE) {
            return res;
        }

        pq = new PriorityQueue<>((a, b) -> b[0] - a[0]);
        pq.offer(new int[]{dist[n - 1], n - 1});
        visited[n - 1] = true;

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int d = current[0];
            int u = current[1];

            for (int[] neighbor : al.get(u)) {
                int v = neighbor[0];
                int e = neighbor[1];
                if (d - edges[e][2] == dist[v]) {
                    res[e] = true;
                    if (!visited[v]) {
                        visited[v] = true;
                        pq.offer(new int[]{dist[v], v});
                    }
                }
            }
        }

        return res;
    }
}
